File: android5/app/src/main/java/repository/PackRepo.kt
/*
 * This file is part of Blokada.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * Copyright © 2022 Blocka AB. All rights reserved.
 *
 * @author Karol Gusak (karol@blocka.net)
 */

package repository

import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import model.*
import service.Services
import ui.PackDataSource
import ui.advanced.packs.convertBlocklists
import utils.Ignored
import utils.Logger
import utils.SimpleTasker
import utils.Tasker


class PackRepo {

    private val api = Services.apiForCurrentUser
    private val dataSource = PackDataSource

    private val cloudRepo by lazy { Repos.cloud }

    private val writeBlocklists = MutableSharedFlow<List<Blocklist>?>(replay = 1)
    private val writePacks = MutableSharedFlow<List<Pack>?>(replay = 1)

    // Blocklists is server side representation of all known lists.
    private val blocklistsHot = writeBlocklists.filterNotNull()

    // Intermediate representation, internal to this class.
    private val mappedBlocklistsHot = blocklistsHot.map { b ->
        convertBlocklists(b.filter { !it.is_allowlist })
    }

    // Used internally to access the map synchronously
    private var mappedBlocklistsInternal = emptyList<MappedBlocklist>()
        @Synchronized set
        @Synchronized get

    // Packs is app's representation. One pack may have multiple configs.
    val packsHot = writePacks.filterNotNull()

    private val loadBlocklistsT = SimpleTasker<Ignored>("loadBlocklists", errorIsMajor = true, timeoutMs = 30000)
    private val convertBlocklistsToPacksT = SimpleTasker<Ignored>("convertBlocklistsToPacks", errorIsMajor = true, timeoutMs = 30000)
    private val installPackT = Tasker<Pack, Ignored>("installPack", errorIsMajor = true, timeoutMs = 30000)
    private val uninstallPackT = Tasker<Pack, Ignored>("uninstallPack", errorIsMajor = true, timeoutMs = 30000)

    fun start() {
        onLoadBlocklists()
        onConvertBlocklistsToPacks()
        onInstallPack()
        onUninstallPack()
        onLoadBlocklists_convertBlocklistsToPacks()
        onBlocklistsIdsChanged_sync()
        onMappedBlocklistsChanged_setField()
    }

    suspend fun installPack(pack: Pack) {
        installPackT.send(pack)
    }

    suspend fun uninstallPack(pack: Pack) {
        uninstallPackT.send(pack)
    }

    // When user changes configuration (selects / deselects) for a pack that is (in)active
    suspend fun changeConfig(pack: Pack, config: PackConfig) {
        val pack = pack.changeStatus(installed = false, config = config)
        installPack(pack)
    }

    fun getPackNameForBlocklist(id: String?): String? {
        val packId = mappedBlocklistsInternal.firstOrNull { it.id == id }?.packId
        return dataSource.getPacks().firstOrNull { it.id == packId }?.meta?.title
    }

    private fun onLoadBlocklists() {
        loadBlocklistsT.setTask {
            val packs = api.getBlocklistsForCurrentUser()
            writeBlocklists.emit(packs)
            it
        }
    }

    // Converts blocklists returned by backend to internal Packs.
    // The app knows a set of Packs (defined in PackDataSource).
    // Here it checks what known packs are active in backend, and ignores the rest.
    private fun onConvertBlocklistsToPacks() {
        convertBlocklistsToPacksT.setTask {
            // Get the intermediate representation, and the backend IDs of active ones.
            val activeBlocklistIds = cloudRepo.blocklistsHot.first()
            val blocklists = mappedBlocklistsHot.first()
            val mapped = blocklists.filter { activeBlocklistIds.contains(it.id) }

            // Map those to the known packs.
            var packs = dataSource.getPacks()
            val packsDict = mutableMapOf<String, Pack>()
            packs.forEach { pack -> packsDict[pack.id] = pack }
            mapped.forEach { mapping ->
                val packId = mapping.packId
                val configName = mapping.packConfig
                val pack = packsDict[packId]

                when {
                    pack == null -> {
                        Logger.w("Pack", "reload: unknown pack id: $packId")
                    }
                    !pack.configs.contains(configName) -> {
                        Logger.w("Pack", "reload: pack $packId doesnt know config $configName")
                    }
                    else -> {
                        val newPack = pack.changeStatus(installed = true, config = configName)
                        packsDict[packId] = newPack
                        packs = packs.map { if (it.id == packId) newPack else it }
                    }
                }
            }

            writePacks.emit(packs)
            true
        }
    }

    private fun onLoadBlocklists_convertBlocklistsToPacks() {
        GlobalScope.launch {
            mappedBlocklistsHot
            .collect {
                convertBlocklistsToPacksT.send()
            }
        }
    }

    private fun onInstallPack() {
        installPackT.setTask { pack ->
            // Select default config for this pack if none selected
            var pack = pack.changeStatus(installing = true, installed = true)
            if (pack.status.config.isEmpty()) {
                Logger.v("Pack", "installPack: selecting first config by default: ${pack.configs.first()}")
                pack = pack.changeStatus(config = pack.configs.first())
            }

            // Announce this pack is installing
            var packs = packsHot.first()
            var newPacks = packs.map { if (it.id == pack.id) pack else it }
            writePacks.emit(newPacks)

            // Get the fresh blocklists information
            val activeBlocklistIds = cloudRepo.blocklistsHot.first()
            val blocklists = mappedBlocklistsHot.first()

            try {
                // Do the actual installation in the Cloud
                val mapped = blocklists.filter {
                    // Get only mapping for selected pack
                    it.packId == pack.id
                            // And only for configs that are active for this pack
                            && pack.status.config.contains(it.packConfig)
                }

                if (mapped.isEmpty()) {
                    throw BlokadaException("Could not find relevant blocklist for: ${pack.id}")
                } else {
                    Logger.v("Pack", "New choice: $mapped")
                }

                // A config might have been unselected for the currently edited pack
                val oldSelectionForThisPack = blocklists.filter { it.packId == pack.id }.map { it.id }

                // Merge lists unique (and maybe deselect a config from current pack)
                val newActiveLists = mapped.map { it.id }.toSet().union(
                    activeBlocklistIds.toSet().minus(oldSelectionForThisPack)
                ).toList()

                cloudRepo.setBlocklists(newActiveLists)

                // Announce this pack is not installing after successful install
                packs = packsHot.first()
                pack = pack.changeStatus(installed = true, updatable = false, installing = false)
                newPacks = packs.map { if(it.id == pack.id) pack else it }
                writePacks.emit(newPacks)
            } catch (ex: Exception) {
                // Announce also if failed installing
                packs = packsHot.first()
                pack = pack.changeStatus(installed = false, installing = false)
                newPacks = packs.map { if(it.id == pack.id) pack else it }
                writePacks.emit(newPacks)
                Logger.e("Packs", "Failed installing pack: ${pack.id}, err: $ex")
            }

            true
        }
    }

    private fun onUninstallPack() {
        uninstallPackT.setTask { pack ->
            // Announce this pack is uninstalling
            var packs = packsHot.first()
            var pack = pack.changeStatus(installing = true, installed = false)
            var newPacks = packs.map { if (it.id == pack.id) pack else it }
            writePacks.emit(newPacks)

            // Get the fresh blocklists information
            val activeBlocklistIds = cloudRepo.blocklistsHot.first()
            val blocklists = mappedBlocklistsHot.first()

            try {
                // Do the actual uninstallation in the Cloud
                val mapped = blocklists.filter {
                    // Get only mapping for selected pack
                    it.packId == pack.id
                }

                if (mapped.isEmpty()) {
                    throw BlokadaException("Could not find relevant blocklist for: ${pack.id}")
                }

                // Merge lists unique
                val newActiveLists = activeBlocklistIds.toSet().minus(mapped.map { it.id }).toList()

                cloudRepo.setBlocklists(newActiveLists)

                // Announce this pack is not uninstalling after successful install
                packs = packsHot.first()
                pack = pack.changeStatus(installed = false, updatable = false, installing = false)
                newPacks = packs.map { if(it.id == pack.id) pack else it }
                writePacks.emit(newPacks)
            } catch (ex: Exception) {
                // Announce also if failed uninstalling
                packs = packsHot.first()
                pack = pack.changeStatus(installed = true, installing = false)
                newPacks = packs.map { if(it.id == pack.id) pack else it }
                writePacks.emit(newPacks)
                Logger.e("Packs", "Failed installing pack: ${pack.id}, err: $ex")
            }
            true
        }
    }

    private fun onBlocklistsIdsChanged_sync() {
        GlobalScope.launch {
            cloudRepo.blocklistsHot
            .collect {
                loadBlocklistsT.send()
            }
        }
    }

    private fun onMappedBlocklistsChanged_setField() {
        GlobalScope.launch {
            mappedBlocklistsHot.collect { mappedBlocklistsInternal = it }
        }
    }

}

File: ios/App/Repository/PackRepo.swift
//
//  This file is part of Blokada.
//
//  This Source Code Form is subject to the terms of the Mozilla Public
//  License, v. 2.0. If a copy of the MPL was not distributed with this
//  file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
//  Copyright © 2021 Blocka AB. All rights reserved.
//
//  @author Karol Gusak
//

import Foundation
import Combine

class PackRepo: Startable {

    // Blocklists is server side representation of all known lists.
    private var blocklistsHot: AnyPublisher<[Blocklist], Never> {
        writeBlocklists.compactMap { $0 }.eraseToAnyPublisher()
    }

    // Intermediate representation, internal to this class.
    private var mappedBlocklistsHot: AnyPublisher<[MappedBlocklist], Never> {
        blocklistsHot.map { it in
            convertBlocklists(blocklists: it.filter { !$0.is_allowlist })
        }.eraseToAnyPublisher()
    }

    // Used internally to access the map synchronously
    private var mappedBlocklistInternal: [MappedBlocklist] = []

    // Packs is app's representation. One pack may have multiple configs.
    var packsHot: AnyPublisher<[Pack], Never> {
        writePacks.compactMap { $0 }.eraseToAnyPublisher()
    }

    private lazy var api = Services.apiForCurrentUser
    private lazy var dataSource = PackDataSource()

    private lazy var cloudRepo = Repos.cloudRepo

    fileprivate let writeBlocklists = CurrentValueSubject<[Blocklist]?, Never>(nil)
    fileprivate let writePacks = CurrentValueSubject<[Pack]?, Never>(nil)

    fileprivate let loadBlocklistsT = SimpleTasker<Ignored>("loadBlocklists", errorIsMajor: true)
    fileprivate let convertBlocklistsToPacksT = SimpleTasker<Ignored>("convertBlocklistsToPacks", errorIsMajor: true)
    fileprivate let installPackT = Tasker<Pack, Ignored>("installPack", errorIsMajor: true)
    fileprivate let uninstallPackT = Tasker<Pack, Ignored>("uninstallPack", errorIsMajor: true)

    private var cancellables = Set<AnyCancellable>()

    func start() {
        onLoadBlocklists()
        onConvertBlocklistsToPacks()
        onInstallPack()
        onUnistallPack()
        onLoadBlocklists_convertBlocklistsToPacks()
        onBlocklistsIdsChanged_sync()
        onMappedBlocklistsChanged_setField()
    }

    func installPack(_ pack: Pack) -> AnyPublisher<Ignored, Error> {
        return installPackT.send(pack)
    }

    func uninstallPack(_ pack: Pack) -> AnyPublisher<Ignored, Error> {
        return uninstallPackT.send(pack)
    }

    // When user changes configuration (selects / deselects) for a pack that is (in)active
    func changeConfig(pack: Pack, config: PackConfig) -> AnyPublisher<Ignored, Error> {
        let pack = pack.changeStatus(installed: false, config: config)
        return installPack(pack)
    }

    func getPackNameForBlocklist(list: String) -> String? {
        let packId = mappedBlocklistInternal.first(where: { $0.id == list })?.packId
        return dataSource.packs.first(where: { $0.id == packId })?.meta.title
    }

    private func onLoadBlocklists() {
        loadBlocklistsT.setTask { _ in Just(true)
            .flatMap { _ in self.api.getBlocklistsForCurrentUser() }
            .map { it in self.writeBlocklists.send(it) }
            .map { _ in true }
            .eraseToAnyPublisher()
        }
    }

    // Converts blocklists returned by backend to internal Packs.
    // The app knows a set of Packs (defined in PackDataSource).
    // Here it checks what known packs are active in backend, and ignores the rest.
    private func onConvertBlocklistsToPacks() {
        convertBlocklistsToPacksT.setTask { _ in Just(true)
            // Get the intermediate representation, and the backend IDs of active ones.
            .flatMap { _ in Publishers.CombineLatest(
                self.cloudRepo.blocklistsHot.first(), self.mappedBlocklistsHot.first()
            )}
            .tryMap { idsAndBlocklists -> [MappedBlocklist] in
                let (activeBlocklistIds, blocklists) = idsAndBlocklists
                return blocklists.filter { it in
                    activeBlocklistIds.contains(it.id)
                }
            }
            // Map those to the known packs.
            .tryMap { mapped -> [Pack] in
                var packs = self.dataSource.packs
                var packsDict: [String: Pack] = [:]
                packs.forEach { pack in packsDict[pack.id] = pack }
                mapped.forEach { mapping in
                    let packId = mapping.packId
                    let configName = mapping.packConfig
                    let pack = packsDict[packId]

                    guard let pack = pack else {
                        return BlockaLogger.w("Pack", "reload: unknown pack id: \(packId)")
                    }

                    guard pack.configs.contains(configName) else {
                        return BlockaLogger.w("Pack", "reload: pack \(packId) doesnt know config \(configName)")
                    }

                    let newPack = pack.changeStatus(installed: true, config: configName)
                    packsDict[packId] = newPack
                    packs = packs.map { $0.id == packId ? newPack : $0 }
                }
                return packs
            }
            .tryMap { it in self.writePacks.send(it) }
            .tryMap { _ in true }
            .eraseToAnyPublisher()
        }
    }

    private func onLoadBlocklists_convertBlocklistsToPacks() {
        mappedBlocklistsHot
        .sink(onValue: { it in self.convertBlocklistsToPacksT.send() })
        .store(in: &cancellables)
    }

    private func onInstallPack() {
        installPackT.setTask { pack -> AnyPublisher<Ignored, Error> in
            // Select default config for this pack if none selected
            var pack = pack.changeStatus(installing: true)
            if pack.status.config.isEmpty {
                BlockaLogger.v("Pack", "installPack: selecting first config by default: \(pack.configs.first!)")
                pack = pack.changeStatus(config: pack.configs.first!)
            }

            return Just(pack)
            // Announce this pack is installing
            .flatMap { it in self.packsHot.first() }
            .tryMap { packs -> Ignored in
                let pack = pack.changeStatus(installing: true)
                let newPacks = packs.map { $0.id == pack.id ? pack : $0 }
                self.writePacks.send(newPacks)
                return true
            }
            // Get the fresh blocklists information
            .flatMap { _ in Publishers.CombineLatest(
                self.cloudRepo.blocklistsHot.first(), self.mappedBlocklistsHot.first()
            )}
            // Do the actual installation in the Cloud
            .tryMap { idsAndBlocklists -> CloudBlocklists in
                let (activeBlocklistIds, blocklists) = idsAndBlocklists

                let mapped = blocklists.filter { it in
                    // Get only mapping for selected pack
                    it.packId == pack.id
                    // And only for configs that are active for this pack
                    && pack.status.config.contains(it.packConfig)
                }

                if mapped.isEmpty {
                    throw "could not find relevant blocklist for: \(pack)"
                } else {
                    BlockaLogger.v("Pack", "New choice: \(mapped)")
                }

                // A config might have been unselected for the currently edited pack
                let oldSelectionForThisPack = blocklists.filter { $0.packId == pack.id }.map { $0.id }

                // Merge lists unique (and maybe deselect a config from current pack)
                let newActiveLists = Set(mapped.map { $0.id }).union(
                    Set(activeBlocklistIds).subtracting(oldSelectionForThisPack)
                )
                return Array(newActiveLists)
            }
            .flatMap { it -> AnyPublisher<Ignored, Error> in self.cloudRepo.setBlocklists(it) }
            // Announce this pack is not installing after successful install
            .flatMap { _ in self.packsHot.first() }
            .tryMap { packs in
                let pack = pack.changeStatus(installed: true, updatable: false, installing: false)
                let newPacks = packs.map { $0.id == pack.id ? pack : $0 }
                self.writePacks.send(newPacks)
                return true
            }
            // Announce also if failed installing
            .tryCatch { err in
                return self.packsHot.first()
                .tryMap { packs -> Ignored in
                    let pack = pack.changeStatus(installed: false, installing: false)
                    let newPacks = packs.map { $0.id == pack.id ? pack : $0 }
                    self.writePacks.send(newPacks)
                    throw err
                }
            }
            .eraseToAnyPublisher()
        }
    }
    
    private func onUnistallPack() {
        uninstallPackT.setTask { pack -> AnyPublisher<Ignored, Error> in return Just(pack)
            // Announce this pack is uninstalling
            .flatMap { it in self.packsHot.first() }
            .tryMap { packs -> Ignored in
                let pack = pack.changeStatus(installing: true)
                let newPacks = packs.map { $0.id == pack.id ? pack : $0 }
                self.writePacks.send(newPacks)
                return true
            }
            // Get the fresh blocklists information
            .flatMap { _ in Publishers.CombineLatest(
                self.cloudRepo.blocklistsHot.first(), self.mappedBlocklistsHot.first()
            )}
            // Do the actual installation in the Cloud
            .tryMap { idsAndBlocklists -> CloudBlocklists in
                let (activeBlocklistIds, blocklists) = idsAndBlocklists
                
                let mapped = blocklists.filter { it in
                    // Get only mapping for selected pack
                    it.packId == pack.id
                }

                if mapped.isEmpty {
                    throw "could not find relevant blocklist for: \(pack)"
                }

                // Merge lists unique
                let newActiveLists = Set(activeBlocklistIds).subtracting(mapped.map { $0.id })
                return Array(newActiveLists)
            }
            .flatMap { it -> AnyPublisher<Ignored, Error> in return self.cloudRepo.setBlocklists(it) }
            // Announce this pack is not installing after successful install
            .flatMap { _ in self.packsHot.first() }
            .tryMap { packs in
                let pack = pack.changeStatus(installed: false, updatable: false, installing: false)
                let newPacks = packs.map { $0.id == pack.id ? pack : $0 }
                self.writePacks.send(newPacks)
                return true
            }
            // Announce also if failed installing
            .tryCatch { err in
                return self.packsHot.first()
                .tryMap { packs -> Ignored in
                    let pack = pack.changeStatus(installed: true, installing: false)
                    let newPacks = packs.map { $0.id == pack.id ? pack : $0 }
                    self.writePacks.send(newPacks)
                    throw err
                }
            }
            .eraseToAnyPublisher()
        }
    }

    private func onBlocklistsIdsChanged_sync() {
        cloudRepo.blocklistsHot
        .sink(onValue: { it in self.loadBlocklistsT.send() })
        .store(in: &cancellables)
    }

    private func onMappedBlocklistsChanged_setField() {
        mappedBlocklistsHot
        .receive(on: RunLoop.main)
        .sink(onValue: { it in self.mappedBlocklistInternal = it })
        .store(in: &cancellables)
    }

}
